lib/src/modules/apply.cc

```diff
@@ -179,6 +182,133 @@ ensure
   end
 end
 
+exit exit_code
+)" };
+
+    // For now just copy it to the cache dir each time unless there is already a copy
+    static const std::string PREP_RUBY_SHIM {
+R"(#! /opt/puppetlabs/puppet/bin/ruby
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'json'
+require 'puppet'
+require 'puppet/configurer'
+require 'securerandom'
+require 'tempfile'
+require 'uri'
+
+## TODO: Option to read from a file is for debugging. Only read from stdin in production.
+args = JSON.parse(ARGV[0] ? File.read(ARGV[0]) : STDIN.read)
+# Create temporary directories for all core Puppet settings so we don't clobber
+# existing state or read from puppet.conf. Also create a temporary modulepath.
+# Additionally include rundir, which gets its own initialization.
+puppet_root = Dir.mktmpdir
+moduledir = File.join(puppet_root, 'modules')
+Dir.mkdir(moduledir)
+plugin_cache = args['plugin_cache']
+plugin_dest = File.join(plugin_cache, 'plugins')
+pluginfact_dest = File.join(plugin_cache, 'pluginfacts')
+# Use isolated directories in puppet_root
+setting_keys = Puppet::Settings::REQUIRED_APP_SETTINGS.append(:rundir)
+cli_base = (setting_keys).flat_map do |setting|
+  ["--#{setting}", File.join(puppet_root, setting.to_s.chomp('dir'))]
+end
+cli_base.concat([
+  '--modulepath',
+  moduledir,
+  '--plugindest',
+  plugin_dest,
+  '--pluginfactdest',
+  pluginfact_dest
+])
+# There will always be at least a single master URI
+# TODO: make sure comma separated is what --server_list expects
+server_list = args['master_uris'].map { |uri| URI.parse(uri).host }.join(',')
+
+cli_settings_pluginsync = [
+  '--localcacert',
+  args['ca'],
+  '--hostcert',
+  args['crt'],
+  '--hostprivkey',
+  args['key'],
+  '--hostcrl',
+  args['crl'],
+  '--server_list',
+  server_list
+]
+
+# Break apart proxy and extract individual settings
+proxy_flags = {
+  user: '--http_proxy_user',
+  password: '--http_proxy_password',
+  port: '--http_proxy_port',
+  host: '--http_proxy_host'
+}
+# Proxy will always be a string (even if its empty)
+parsed_proxy = URI.parse(args['proxy'])
+proxy_flags.each do |uri_method, setting_flag|
+  val = parsed_proxy.send(uri_method)
+  cli_settings_pluginsync << setting_flag << val unless val.nil?
+end
+
+exit_code = 0
+begin
+
```

Comment:
Once we pull these out into shipped files we can look in to sharing some code between `prep` and `apply`.

Reply:
It might make more sense to have one shim that can take the action to perform as a param.

Reply:
For sure. Updated to a single shim. I started the work thinking they would be pretty different, turns out they are not.